Easier Command Based Programming

FRC2539 opened the pull request:
This is the C++ side of issue #235. It implements default life-cycle methods in the `Command` class, adds two new classes (`TimedCommand` and `InstantCommand`) to handle common behaviors, and updates the existing commands to take advantage of these.

frcjenkins commented:
Can one of the admins verify this patch?

calcmogul commented on wpilibc/shared/include/Commands/InstantCommand.h:
```diff
@@ -0,0 +1,20 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) FIRST 2011-2016. All Rights Reserved.                        */
```
Since these files are new, the copyright year should only be "2016".

calcmogul commented on wpilibc/shared/src/Commands/InstantCommand.cpp:
```diff
@@ -0,0 +1,15 @@
+
+InstantCommand::InstantCommand(const std::string& name) : Command(name) {}
+
+InstantCommand::InstantCommand() : Command() {}
```
Since only the default base class constructor is called and the constructor body is empty, the header can have `InstantCommand() = default;` and this line can be removed.

calcmogul commented on wpilibc/shared/src/Commands/TimedCommand.cpp:
```diff
@@ -0,0 +1,14 @@
+
+#include "Commands/TimedCommand.h"
+
+TimedCommand::TimedCommand(const std::string& name, double timeout) : Command(name, timeout) {}
```
format.py should be run. It would wrap lines like this.

calcmogul commented on wpilibc/shared/src/Commands/WaitCommand.cpp:
```diff
@@ -9,21 +9,11 @@
 WaitCommand::WaitCommand(double timeout)
-    : Command(
+    : TimedCommand(
           ((std::stringstream&)(std::stringstream("Wait(") << timeout << ")"))
```
Don't use C-style casts.

calcmogul replied:
Running that line locally returned `5)it(`. `("Wait(" + std::to_string(timeout) + ")").c_str()` produces `Wait(5.000000)`, which is probably the output you intended.

calcmogul commented on wpilibc/shared/src/Commands/WaitCommand.cpp:
```diff
@@ -9,11 +9,7 @@
 #include <sstream>
```
Since stringstream is no longer used,  this include can be removed.

calcmogul commented on wpilibc/shared/include/Commands/TimedCommand.h:
```diff
@@ -1,20 +1,21 @@
 
 #pragma once
 
+#include <string>
```
Per our style guide, there should be an empty line between header types. In this case, between C++ system includes and user includes.

FRC2539 replied:
Interestingly enough, format.py suggested adding the `<string>` include, and then moved it directly above the other one (we had placed it below).

calcmogul replied:
The formatter currently enforces the correct order, but doesn't yet add newlines between groups. That's on my to do list after I finish some documentation changes and unit tests.

calcmogul commented on wpilibc/shared/include/Commands/InstantCommand.h:
```diff
@@ -1,20 +1,21 @@
 
 #pragma once
 
+#include <string>
```
Add an empty line between these. See the other comment.

calcmogul commented:
Here's a patch for the Java implementation: https://paste.unixcube.org/t/5f8a82

FRC2539 commented:
Running format.py on PrintCommand.cpp in master changed `"stdio.h"` to `<cstdio>`. PrintCommand.cpp in this PR does not use `"stdio.h"` or `<cstdio>`, so presumably format.py would not have made any changes to it. So the correct resolution should be to overwrite PrintCommand.cpp with the version in this PR.

Do you need us to resolve the conflict? We're not sure how to rebase our changes since we did not create a branch for this PR.

calcmogul commented:
Since there are a bunch of small commits that conflict, rebasing this as-is looks tricky (I tried it). I recommend squashing the changes first with `git rebase -i <commit hash preceding your commits>` by setting the later commits to "fixup".

After that, run `git pull --rebase git://github.com/wpilibsuite/allwpilib master` to fetch the new commits and rebase your local changes on top of them. Then fix the conflicts.

FRC2539 commented:
We were able to use the [GitHub instructions for syncing a fork](https://help.github.com/articles/syncing-a-fork/) to rebase our changes. There were only two conflicts, both within `PrintCommand.cpp`, so it wasn't as hard as we had been expecting. Is there anything we can do to help get this merged before new conflicts happen?

FRC2539 commented:
Is there anything we can do to help get this merged before there are new conflicts?

calcmogul commented on wpilibc/shared/include/Commands/TimedCommand.h:
```diff
@@ -0,0 +1,22 @@
+
+class TimedCommand : public Command {
+ public:
+  explicit TimedCommand(const std::string& name, double timeout);
```
Since this constructor has two arguments, it shouldn't be marked explicit.

FRC2539 commented:
Thank for the insight. Waiting has helped uncover a few issues that might have gotten through otherwise, so it make sense. We've squashed all the commits into one, and added the issue number to the commit message.

calcmogul commented:
The number after the commit message is supposed to be the PR number, not the issue number. If the intent is to close the issue when this commit is merged, then add "Closes #235" to the commit message. See https://help.github.com/articles/closing-issues-via-commit-messages/.

FRC2539 commented:
We'll amend the commit message the next time we have to make changes, or right before you're ready to merge if no more changes are needed. Feel free to have more people review it in the meantime.

AustinShalit commented:
@frcjenkins retest this please

333fred commented:
@frcjenkins retest this please

calcmogul commented:
This needs to be rebased against master and tested again. The issue causing compilation to fail has been fixed.

bradamiller commented:
@frcjenkins retest this please

calcmogul commented:
@FRC2539 could you rebase this against master please?

frcjenkins commented:
Can one of the admins verify this patch?

frcjenkins commented:
Can one of the admins verify this patch?

calcmogul commented on wpilibj/src/shared/java/edu/wpi/first/wpilibj/command/InstantCommand.java:
```diff
@@ -0,0 +1,22 @@
+
+public class InstantCommand extends Command {
+  public InstantCommand() {
+    super();
```
Doesn't the default constructor get called automatically in this case? Calling `super()` explicitly shouldn't be necessary.

FRC2539 replied:
We don't know much about Java, especially how its inheritance works. We just used the [patch](https://paste.unixcube.org/t/5f8a82) you linked from pastebin. If you think it can be better, tell us what to change and we'll do it.

calcmogul replied:
I'd forgotten that... I successfully tested removing the "super();" line, so you can go ahead and do that.

FRC2539 replied:
This is probably a stupid question, but should we get rid of the now empty constructor?

PeterJohnson commented:
@frcjenkins retest this please
